### PR TITLE
[AutoSP] Fix compiler correctness

### PR DESCRIPTION
**Pull request by jinyouzhi**
## Summary

This PR fixes correctness and robustness issues in the AutoSP compiler pass (Ulysses sequence parallelism). The previous implementation could silently produce incorrect losses, rewrite tensors along the wrong dimension, or fail inside a collective when model layouts did not match its implicit assumptions.

### 1. Causal-LM loss was incorrect at sequence-shard boundaries

For a causal language-model loss, the target for token `t` is commonly formed by shifting labels by one position. If labels are sliced independently on each rank before the shift, the first/last token at every shard boundary is paired with the wrong logit or dropped. With sequence length 8 and `sequence_parallel_size=2`, the target crossing the boundary between positions 3 and 4 is not represented by either local shard. A local mean reduction also gives the wrong global loss when ranks have different numbers of valid (`ignore_index`) tokens.

The fix detects supported shifted-label paths, shards the shifted target, counts valid tokens per rank, and aggregates the loss using a valid-token-weighted mean with an autograd-aware custom op. Unsupported class-weighted or non-mean causal losses now fail explicitly instead of producing an incorrect result; ordinary non-causal label paths continue to use normal label sharding.

### 2. Non-default sequence dimensions were interpreted as dimension 1

Models using sequence-first inputs, for example `[S, B]`, were previously sliced along the batch dimension because AutoSP assumed `seq_dim=1`. This can result in incorrect input/label/position-id shapes and downstream attention failures. The selected `seq_dim` is now stored in the compiler metadata and used consistently for all sharding and symbolic slicing operations.

### 3. Invalid sequence/head layouts reached reshape or collective operations

AutoSP requires the global sequence length and every SDPA Q/K/V head count to be divisible by `sequence_parallel_size`. Previously, a non-divisible value could reach `reshape`/`all_to_all`, causing a shape error or mismatched collective behavior. The compiler now validates the mesh, sequence divisibility, tensor rank, and head divisibility, and reports the violated requirement directly.

### 4. Attention masks were left with local query dimensions

After Q/K/V exchange, SDPA operates on a global sequence dimension. A mask whose query dimension was still local could therefore broadcast incorrectly or fail with a size mismatch. Supported masks are reconstructed by gathering the query dimension before SDPA. Masks that require gathering a sharded key dimension are rejected with an actionable error because they cannot be reconstructed safely from local inputs.

### 5. AutoSP process groups could become stale across compilations

The process-group registry previously reused an initialized mesh even when a later compilation requested a different SP/DP layout. That could pair the graph with the wrong communicator. The requested mesh must now cover the world size exactly, and changing an already-initialized mesh raises an explicit error rather than silently reusing stale groups.

### 6. Fake shape propagation and GQA backward shapes were fragile

Torch 2.9 fake propagation can fail on SDPA masks even though AutoSP only needs output shape metadata; the pass now temporarily removes and restores masks during shape propagation. FPDT backward now allocates K/V gradient buffers from the K/V shapes, which fixes grouped-query attention cases where K/V head counts differ from Q.

## Tests

- `pytest -q tests/unit/v1/compile/test_compile_autosp.py -k AutoSPValidation`
- Targeted sequential AutoSP graph-rewrite tests covering symbolic slicing, sequence-first layouts, causal-label handling, mask reconstruction, mesh validation, and head/sequence divisibility checks
- Distributed AutoSP coverage across `sequence_parallel_size=2` and `4`, ZeRO stages 0 and 1, and FP32/BF16 configurations where supported
- GitHub Actions: `modal-torch-latest / collect tests` passed
- Changed-file pre-commit hooks passed

Verified locally against PR base 1b892a75 and PR head 1578435cd: the 13 AutoSPValidation regression tests for scenarios 1-5 all failed on the base and passed on the PR head. Scenario 1 is validated at the graph-rewrite and valid-token aggregation level, not by a numerical cross-shard loss comparison. Scenario 6 (Torch 2.9 fake shape propagation and FPDT GQA backward buffer shapes) has no corresponding new UT in this PR; it is covered by code-path review and existing broader tests rather than a before/after regression test.

The expected behavior for unsupported layouts is now an explicit validation error; supported layouts preserve the original loss semantics and attention-mask behavior while distributing the sequence dimension.

### CODE REVIEW

File: deepspeed/compile/custom_ops/all_to_all.py [hunk 1]
```diff
@@ -90,3 +93,58 @@ def _all_to_all_backward(ctx, grad):
...
+
+@torch.library.register_fake("autosp::all_gather_sequence")
+def all_gather_sequence_fake(input: torch.Tensor, dim: int):
+    output_shape = list(input.shape)
+    output_shape[dim] *= sp_size()
+    return input.new_empty(output_shape)
+
+
+@torch.library.custom_op("autosp::aggregate_loss", mutates_args=())
+def aggregate_loss(loss: torch.Tensor, valid_tokens: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (jinyouzhi) [reply deepspeed/compile/custom_ops/all_to_all.py]: Fixed

File: deepspeed/compile/passes/sp_compile.py [hunk 2]
```diff
@@ -133,7 +134,82 @@ def pass_shard_input_ids(gm: GraphModule, example_inputs):
...
+                current_seq_len = current_meta.shape[seq_dim]
+                if str(current_seq_len) == str(seq_len):
+                    shifted_label_node = current
+                    break
+            worklist.extend(current.all_input_nodes)
+
+        if shifted_label_node is None:
+            raise RuntimeError("AutoSP causal language-model loss requires labels to be shifted to the original "
+                               "sequence length before flattening. Pass pre-shifted labels or use a Transformers "
+                               "causal loss that pads labels before shifting.")
```
[the remark this answers is not in the document]
Comment (jinyouzhi) [reply deepspeed/compile/passes/sp_compile.py]: Fixed.

**Comment by pengdurice** [comment]
thanks for the PR, will take a closer look later, but can you help add more details of the issues, your fixes, and the test results.

**Comment by jinyouzhi** [comment]
> thanks for the PR, will take a closer look later, but can you help add more details of the issues, your fixes, and the test results. 

Sure, thank you. Will add more details asap.
